### DISCUSSION TITLE
Clarification on using blank detection and fallback on video sources

### DISCUSSION THREAD

**Discussion by madhermit** [savonet/liquidsoap#3191] [Q&A]
Hello! I'm trying to create a script that takes a video sources as input, and switches to a static image when it detects issues with the  input stream, and switches back when the stream has recovered.

So far I have the following:

```
#!/usr/bin/env liquidsoap
log.level.set(4)

input_url ="udp://0.0.0.0:1938"
output_url="udp://0.0.0.0:8000"

stream = input.ffmpeg(input_url)
interlude = mksafe(image(fallible=false, "failover.png"))

## Attempt #1:
## GOOD - if script started with no input stream, it correctly plays interlude,
## GOOD - if input stream started later, then correctly switches/plays the input stream
## BAD  - if input stream killed during playback, playback dies -- it should switch to interlude
# stream = fallback(track_sensitive=false, [ stream, interlude ])

## Attempt #2:
## GOOD - if script started with no input stream, it correctly plays interlude,
## GOOD - if input stream started later, then correctly switches/plays the input stream
## BAD  - if input stream killed during playback, playback dies -- it should switch to interlude
stream = fallback(track_sensitive=false, [ blank.strip(track_sensitive=false, max_blank=5., stream) , interlude ])

enc = %ffmpeg(format="mpegts", %video(codec="libx264"), %audio(codec="mp2"))
output.url(enc, self_sync=true, url=output_url, stream)
```

I'm setting up the input stream using ffmpeg and then viewing the output stream using vlc. As mentioned in the comments, if i start the liquidsoap script without the input stream running, it will display the interlude correctly; if I then start the input stream, the liquidsoap script will detect it and stream it; this is all good so far. If then I stop the input stream (kill the ffmpeg process), the liquidsoap script does not detect this and the stream play stalls. I want it to detect this and switch to the interlude fallback.

Am I using/understanding the blank and/or fallback detection correctly here with a video source? If so then how can I modify the above script to handle the failover/fallback correctly?

Thanks in advance!

**Comment by toots** [comment]
Hi! Thanks for reporting!

`blank` operators are designed to work with audio only. I realize now that the name is highly confusing now that we do support video, they are historical..

If your concern is that the stream may stop working then `fallback` operators should suffice:

```liquidsoap
log.level.set(4)

input_url ="udp://0.0.0.0:1938"
output_url="udp://0.0.0.0:8000"

stream = input.ffmpeg(input_url)
interlude = mksafe(image(fallible=false, "failover.png"))

stream = fallback(track_sensitive=false, [ stream, interlude ])

enc = %ffmpeg(format="mpegts", %video(codec="libx264"), %audio(codec="mp2"))
output.url(enc, self_sync=true, url=output_url, stream)
```

It seems to me that the issue could come from the fact that `input.ffmpeg` fails to detect that the input stream is not delivering data. Do you have some logs to share? Make sure to set `log.level.set(4)`. 

Thanks!

**Comment by Moonbase59** [comment]
Heretic side thought: Maybe `blank.strip` & `blank.eat` could be expanded to work on video, too? Like detecting video _and_ audio being "blank"? Just a thought, can’t see a real use case…

**Comment by vitoyucepi** [comment]
As I noticed previously in the other [discussion](https://github.com/savonet/liquidsoap/discussions/3167#discussioncomment-6259066).
> Some will work better than others.

```ruby
a = input.rtmp("rtmp://0.0.0.0:1935/a/b")
#a = input.srt(bind_address="0.0.0.0", port=1936)
#a = input.harbor("/live", port=1937)
#a = input.ffmpeg("udp://0.0.0.0:1938")

video = single("image.jpg")

s = fallback([a, video], track_sensitive=false)

enc = %ffmpeg(format="mpegts", %video(codec="libx264"))

output.harbor.hls([("test", enc)], s)
```

To view the result use `ffplay http://localhost:8000/test.m3u8` or `streamlink http://localhost:8000/test.m3u8 best`.

I think that `a = input.ffmpeg("udp://0.0.0.0:1938")` is not reliably implemented at the moment. Occasionally, I'm unsure why, but liquidsoap may not terminate normally.

**Reply by toots** [reply]
Yes. udp might be the problem. It's pretty low-level with no notion of connection status/timeout so that's a possibility.
